ops/model.py

```diff
@@ -810,6 +895,281 @@ def __init__(self, name: str, address_info: '_AddressDict'):
+                'expires={self.expires!r}, '
+                'rotation={self.rotation}, '
+                'rotates={self.rotates!r})'
+                ).format(self=self)
+
+
+class Secret:
+    """Represents a single secret in the model.
+
+    This class should not be instantiated directly, instead use
+    :meth:`Model.get_secret` (for consumers and owners), or
+    :meth:`Application.add_secret` or :meth:`Unit.add_secret` (for owners).
+
+    All secret events have a :code:`.secret` attribute which provides the
+    :class:`Secret` associated with that event.
+    """
+
+    _key_re = re.compile(r'^([a-z](?:-?[a-z0-9]){2,})$')  # copied from Juju code
+
+    def __init__(self, backend: '_ModelBackend',
+                 id: Optional[str] = None,
+                 label: Optional[str] = None,
+                 content: Optional[Dict[str, str]] = None):
+        if not (id or label):
+            raise TypeError('Must provide an id or label, or both')
+        if id is not None:
+            id = self._canonicalize_id(id)
+        self._backend = backend
+        self._id = id
+        self._label = label
+        self._content = content
+
+    def __repr__(self):
+        fields = []  # type: List[str]
+        if self._id is not None:
+            fields.append('id={!r}'.format(self._id))
+        if self._label is not None:
+            fields.append('label={!r}'.format(self._label))
+        return '<Secret ' + ' '.join(fields) + '>'
+
+    @classmethod
+    def _canonicalize_id(cls, id: str) -> str:
+        """Return the canonical form of the given secret ID, with the 'secret:' prefix."""
+        id = id.strip()
+        if not id.startswith('secret:'):
+            id = 'secret:' + id  # add the prefix if not there already
+        return id
+
+    @classmethod
+    def _validate_content(cls, content: Optional[Dict[str, str]]):
+        """Ensure the given secret content is valid, or raise ValueError."""
+        if not content:
+            raise ValueError('Secret content must not be empty')
+
+        invalid_keys = [k for k in content if not cls._key_re.match(k)]
+        invalid_values = [v for v in content.values() if not isinstance(v, str)]  # type: List[Any]
+        msg = ''
+        if invalid_keys:
+            msg += ('Invalid secret keys: {}. Keys should be at least 3 characters long, '
+                    + 'start with a letter, and not end with a hyphen.').format(invalid_keys)
+        if invalid_values:
+            if msg:
+                msg += ' '
+            type_names = [type(v).__name__ for v in invalid_values]
+            msg += 'Invalid secret values: should be of type str, not {}.'.format(
+                ' or '.join(type_names))
+        if msg:
+            raise ValueError(msg)
+
+    @property
+    def id(self) -> Optional[str]:
+        """Unique identifier for this secret.
+
+        This will be None if you obtained the secret using
+        :meth:`Model.get_secret` with a label but no ID.
```

Comment:
Yeah, Juju doesn't give an ID when you ask for the secret content (`secret-get`). The owner can get it when getting metadata via via `secret-get-info`, but not the consumer. I think it would have been good if `secret-get` returned this, but oh well.

Reply:
Per discussion, this feels like an oversight that as a consumer if you have a label you can't find the ID. I'll raise with Ian and co.

Reply:
We're going to punt on this and can discuss later if there's a real-world use case.